src/us/kbase/workspace/kbase/LocalTypeServerMethods.java

```diff
@@ -0,0 +1,323 @@
+import us.kbase.workspace.ListModuleVersionsParams;
+import us.kbase.workspace.ListModulesParams;
+import us.kbase.workspace.ModuleInfo;
+import us.kbase.workspace.ModuleVersions;
+import us.kbase.workspace.RegisterTypespecCopyParams;
+import us.kbase.workspace.RegisterTypespecParams;
+import us.kbase.workspace.RemoveModuleOwnershipParams;
+import us.kbase.workspace.TypeInfo;
+import us.kbase.workspace.WorkspaceClient;
+import us.kbase.workspace.database.Types;
+import us.kbase.workspace.database.WorkspaceUser;
+
+public class LocalTypeServerMethods {
+	
+	// TODO TEST unit tests, covered by integration tests for now
+	// TODO JAVADOC
+	
+	private final Types types;
+
+	public LocalTypeServerMethods(final Types types) {
+		this.types = types;
+	}
+	
+	public Types getTypes() {
+		return types;
+	}
+
+	private WorkspaceUser getUser(final AuthToken token) {
+		return token == null ? null : new WorkspaceUser(token.getUserName());
+	}
+	
+	public void grantModuleOwnership(
+			final GrantModuleOwnershipParams params,
+			final AuthToken authPart,
+			final boolean asAdmin)
+			throws TypeStorageException, NoSuchPrivilegeException {
+		checkAddlArgs(params.getAdditionalProperties(), GrantModuleOwnershipParams.class);
+		types.grantModuleOwnership(params.getMod(), params.getNewOwner(),
+				longToBoolean(params.getWithGrantOption()), getUser(authPart), asAdmin);
+	}
+	
+	public void removeModuleOwnership(
+			final RemoveModuleOwnershipParams params,
+			final AuthToken authPart,
+			final boolean asAdmin)
+			throws NoSuchPrivilegeException, TypeStorageException {
+		checkAddlArgs(params.getAdditionalProperties(), RemoveModuleOwnershipParams.class);
+		types.removeModuleOwnership(params.getMod(), params.getOldOwner(),
+				getUser(authPart), asAdmin);
+	}
+
+	public void requestModuleOwnership(final String mod, final AuthToken authPart)
+			throws TypeStorageException {
+		types.requestModuleRegistration(getUser(authPart), mod);
+	}
+	
+	public Map<String,String> registerTypespec(
+			final RegisterTypespecParams params,
+			final AuthToken authPart)
+			throws SpecParseException, TypeStorageException, NoSuchPrivilegeException,
+				NoSuchModuleException {
+		//TODO improve parse errors, don't need include path, currentlyCompiled
+		checkAddlArgs(params.getAdditionalProperties(), params.getClass());
+		if (!(params.getMod() == null) ^ (params.getSpec() == null)) {
+			throw new IllegalArgumentException(
+					"Must provide either a spec or module name");
+		}
+		final List<String> add = params.getNewTypes() != null ?
+				params.getNewTypes() : new ArrayList<String>();
+		final List<String> rem = params.getRemoveTypes() != null ?
+				params.getRemoveTypes() : new ArrayList<String>();
+		final Map<String, Long> deps = params.getDependencies() != null ?
+				params.getDependencies() : new HashMap<String, Long>();
+		final Map<TypeDefName, TypeChange> res;
+		if (params.getMod() != null) {
+			 res = types.compileTypeSpec(
```

Comment:
is this going to be covered somewhere else / in a future PR?

Reply:
Not in the near future anyway. It hasn't been covered for the last 8 years apparently, so for now I'm leaving things as is.